test/blackbox/common/DDSBlackboxTestsDataWriter.cpp

```diff
@@ -1296,6 +1296,248 @@ TEST_P(DDSDataWriter, datawriter_sends_non_default_qos_optional)
+}
+
+/**
+ * @test DataWriter Sample prefilter feature
+ *
+ * This test asserts that prefiltering by payload on a DataWriter
+ * correctly works. The test sets a prefilter to only send samples
+ * with index >= 5 to the discovered reader.
+ */
+TEST_P(DDSDataWriter, datawriter_prefilter_filtering_by_payload)
+{
+    // TODO(Mario-DL): Remove when multiple filtering readers case is fixed for data-sharing
+    if (enable_datasharing)
+    {
+        GTEST_SKIP() << "Several filtering readers not correctly working on data sharing";
+    }
+
+    struct CustomUserWriteData : public rtps::WriteParams::UserWriteData
+    {
+        CustomUserWriteData(
+                const uint16_t& desired_idx)
+            : desired_idx_(desired_idx)
+        {
+        }
+
+        uint16_t desired_idx_;
+    };
+
+    struct CustomPreFilter : public eprosima::fastdds::dds::IContentFilter
+    {
+        ~CustomPreFilter() override = default;
+
+        //! Custom filter for the HelloWorld example
+        bool evaluate(
+                const SerializedPayload& payload,
+                const FilterSampleInfo& filter_sample_info,
+                const rtps::GUID_t&) const override
+        {
+            HelloWorldPubSubType hello_world_type_support;
+            HelloWorld hello_world_sample;
+
+            hello_world_type_support.deserialize(*const_cast<SerializedPayload*>(&payload), &hello_world_sample);
+
+            bool sample_should_be_sent = true;
+
+            auto custom_write_data =
+                    std::static_pointer_cast<CustomUserWriteData>(filter_sample_info.user_write_data);
+
+            // Filter out samples <= desired_index
+            if (hello_world_sample.index() <= custom_write_data->desired_idx_)
+            {
+                sample_should_be_sent = false;
+            }
+            return sample_should_be_sent;
+        }
+
+    };
+
+    PubSubWriter<HelloWorldPubSubType> writer(TEST_TOPIC_NAME);
+    PubSubReader<HelloWorldPubSubType> reader(TEST_TOPIC_NAME);
+
+    // Initialize writer and the filtered reader
+    writer.init();
+    ASSERT_TRUE(writer.isInitialized());
+    reader.init();
+    ASSERT_TRUE(reader.isInitialized());
+
+    // wait for discovery between writer and filtered reader
+    writer.wait_discovery();
+    reader.wait_discovery();
+
+    // Set a prefilter on the filtered reader
+    ASSERT_EQ(writer.set_sample_prefilter(
+                std::make_shared<CustomPreFilter>()),
+            eprosima::fastdds::dds::RETCODE_OK);
+
+    // Set a user write data on the writer to filter out samples with idx <= 5
+    rtps::WriteParams write_params;
+    write_params.user_write_data(std::make_shared<CustomUserWriteData>(
+                (uint16_t)5u));
+
+    auto data = default_helloworld_data_generator();
+
+    reader.startReception(data);
+
+    writer.send(data, 50, &write_params);
+    // Reader should have received the samples
+    ASSERT_EQ(reader.block_for_all(std::chrono::seconds(1)), 5u);
+}
+
```

Comment:
We need a test that checks that the prefilter works when the standard filtering is disabled (i.e. `reader_filters_allocation.maximum = 0`)